### ISSUE TITLE
Towards a package

### ISSUE THREAD

**Issue by ocots** [control-toolbox/control-loss#14]
We could imagine to make a package to solve optimal control problems with regions of loss control.

First, we need to automatise the direct part. This may be done in several steps. First, we need a modelisation of the optimal control problem with loss control region. Then, we need an algorithm with transform the problem into a classical optimal control problem that can be solved by OptimalControl.jl. Note that the modelisation may be helped also by OptimalControl.jl. At the end the resolution is done by OptimalControl.jl but there will be maybe some parameters to tune or homotopic parameters for instance for the regularisation. The $\varepsilon$ parameter may be fixed in the algorithm or updated.

We can use this issue to share ideas.

**Comment by ocots** [comment]
@AnasXbouali Can you write here the general definition of an optimal control problem with loss control regions?

**Comment by ocots** [comment]
I think that we should write in this issue how a user would interact with the package to solve [this example](https://control-toolbox.org/control-loss/stable/zermelo1.html). The transformation to a classical problem is internal and the user won't see it.

The key point is how to define the loss control regions.

First attempt difficult to generalize:

```julia
tf  = 8
ocp = @def begin

    t ∈ [ 0, tf ], time
    x ∈ R^2, state
    u ∈ R, control

    0.5 ≤  x₂(t)  ≤ 3.5, loss_region

    x₁(0) == 0
    x₂(0) == 0
    x₂(tf) == 4

    -π/2  ≤  u(t)  ≤ π/2

    ẋ(t) == [ x₂(t) + cos(u(t)), sin(u(t)) ]

    -x₁(tf) → min

end
```

**Comment by ocots** [comment]
We should write some examples of definitions of loss regions. Below, each line is a region:

```julia
 @def begin

2 ≤ x₂(t) ≤ 3, loss_region
0 ≤ x₁(t)^2 + x₂(t)^2 ≤ 1, loss_region
( x₁(t), x₂(t) ) ∈ [0, 1] × [2, 3], loss_region
0 ≤ x₁(t)^2 ≤ 1 and 2 ≤ x₂(t) ≤ 3, loss_region

end
```

**Comment by ocots** [comment]
## Important remark

Since it is not so simple to parse the code which defines the loss control regions, we can in a first attempt make something like that:

```julia
tf  = 8
ocp = @def begin

    t ∈ [ 0, tf ], time
    x ∈ R^2, state
    u ∈ R, control

    x₁(0) == 0
    x₂(0) == 0
    x₂(tf) == 4

    -π/2  ≤  u(t)  ≤ π/2

    ẋ(t) == [ x₂(t) + cos(u(t)), sin(u(t)) ]

    -x₁(tf) → min

end

loss_regions!(ocp; rg = 2, lb = 2, ub = 3)

sol = solve(ocp; ε₀ = 1e-3, ε = 1e-3)

plot(sol)
```

Voir [constraint!](https://github.com/control-toolbox/CTBase.jl/blob/a9744ff7cfaed4126c2bdbddce55b8adbf344010/src/optimal_control_model-setters.jl#L520).

**Comment by ocots** [comment]
Let split the methods:

```julia
# definition of a classical ocp

tf  = 8
ocp = @def begin

    t ∈ [ 0, tf ], time
    x ∈ R^2, state
    u ∈ R, control

    x₁(0) == 0
    x₂(0) == 0
    x₂(tf) == 4

    -π/2  ≤  u(t)  ≤ π/2

    ẋ(t) == [ x₂(t) + cos(u(t)), sin(u(t)) ]

    -x₁(tf) → min

end
```

```julia
# switch to problem with loss control
loss_ocp = LossControlProblem(ocp)

# add loss control regions
loss_regions!(loss_ocp; rg = 2, lb = 2, ub = 3)
```

```julia
# convert loss control problem to a classical one
ocp = OptimalControlProblem(loss_ocp; ε₀ = 1e-3, ε = 1e-3))
```

```julia
# solve the classical problem
sol = solve(ocp)
```

```julia
# plot the classical solution
plot(sol)

# or plot a solution from a loss control problem
sol_loss = LossControlSolution(sol, ocp_loss)
plot(sol_loss)
```

**Comment by ocots** [comment]
In the previous comment, `LossControlModel` is used to get a struct that can save loss control regions. The function `loss_regions` permits to add loss control regions. The function `OptimalControlModel` is the method which regularize the problem, add the variable, etc.

**Comment by ocots** [comment]
En fait, ce serait mieux de rester au niveau abstrait.

**Comment by AnasXbouali** [comment]
D'accord, je vais essayer de travailler sur cette transformation pour l'exemple de Zermelo 1. Ensuite, nous verrons les aspects qui peuvent être automatisés.

**Comment by ocots** [comment]
Ce serait en effet le Graal de passer directement par l'abstrait mais cela me semble assez difficile. Je reviens donc sur ce que j'ai dit et il me semble que suivre dans un premier temps ceci https://github.com/control-toolbox/LossControl.jl/issues/14#issuecomment-2334126189 est plus simple. Un compromis serait la chose suivante.

On part de : 

```julia
tf  = 8
ocp = @loss_def begin

    t ∈ [ 0, tf ], time
    x ∈ R^2, state
    u ∈ R, control

    0.5 ≤  x₂(t)  ≤ 3.5, loss_region

    x₁(0) == 0
    x₂(0) == 0
    x₂(tf) == 4

    -π/2  ≤  u(t)  ≤ π/2

    ∂x₁ = x₂(t) + cos(u(t))
    ∂x₂ = sin(u(t))

    ẋ(t) == [ ∂x₁, ∂x₂ ]

    -x₁(tf) → min

end
```

J'ai fait exprès j'ai écrit la fonction de parsing `@loss_def`. 

- On parse tout d'abord avec `@def` de `CTBase.jl` en enlevant au préalable toutes les lignes contenant l'étiquette `loss_region`. Voir [ici](https://github.com/control-toolbox/CTBase.jl/blob/a9744ff7cfaed4126c2bdbddce55b8adbf344010/src/onepass.jl#L77) pour un exemple de parsing.
- On ajoute autant d'états qu'il y a de contrôle et on ajoute les contraintes associées. Voir les fichiers [OptimalControlModel](https://github.com/control-toolbox/CTBase.jl/blob/main/src/optimal_control_model-type.jl), les [getters](https://github.com/control-toolbox/CTBase.jl/blob/main/src/optimal_control_model-getters.jl) et les [setters](https://github.com/control-toolbox/CTBase.jl/blob/main/src/optimal_control_model-setters.jl).
- On ajoute le contrôle permettant de changer de valeur entre les régions de perte de contrôle.
- On parse et parcours les régions de perte de contrôle indiquées par l'étiquette `loss_region` et on met à jour la fonction indicatrice régularisée. On met aussi à jour la dynamique à chaque fois que l'on ajoute une région de perte de contrôle.
- On met à jour le critère.

>[!NOTE]
> Il faut une structure de donnée permettant de stocker les régions de perte de contrôle, la fonction indicatrice, l'ocp...
